src/pstl_offload.cpp

```diff
@@ -29,12 +29,54 @@
 namespace __pstl_offload
 {
 
-static auto
-__get_original_free()
+using __free_func_type = void (*)(void*);
+
+// list of objects for delayed releasing
+struct __delayed_free_list {
+    __delayed_free_list* _M_next;
+    void*                  _M_to_free;
+};
+
+// are we inside dlsym call?
+static thread_local bool __dlsym_called = false;
+// objects released inside of dlsym call
+static thread_local __delayed_free_list* __delayed_free = nullptr;
+
+static __free_func_type
+__get_original_free_checked(void* __ptr_to_free)
+{
+    __dlsym_called = true;
+    __free_func_type __orig_free = __free_func_type(dlsym(RTLD_NEXT, "free"));
+    __dlsym_called = false;
+    if (!__orig_free)
+    {
+        throw std::system_error(std::error_code(), dlerror());
+    }
+
+    // Releasing objects from delayed release list. It's enough to check __delayed_free
+    // only at this point, as __delayed_free filled only inside dlsym(RTLD_NEXT, "free").
+    while (__delayed_free)
+    {
+        __delayed_free_list* __next = __delayed_free->_M_next;
+        // it's possible that an object to be released during 1st call of __internal_free
+        // would be released 2nd time from inside nested dlsym call. To prevent "double free"
+        // situation, check for it explicitly.
+        if (__ptr_to_free != __delayed_free->_M_to_free)
+        {
+            __orig_free(__delayed_free->_M_to_free);
+        }
+        __orig_free(__delayed_free);
+        __delayed_free = __next;
+    }
+
+    return __orig_free;
+}
+
+static __free_func_type
+__get_original_free(void* __ptr_to_free)
 {
-    using __free_func_type = void (*)(void*);
+    static __free_func_type __orig_free = __get_original_free_checked(__ptr_to_free);
```

Comment:
May be it also makes sense to immediately release the `__ptr_to_free` and rename the function:

```suggestion
    static __free_func_type __orig_free = __get_original_free_checked(__ptr_to_free);
    __orig_free(__ptr_to_free);
```
It will allow to avoid passing this pointer twice on the calling side.

Reply:
You are suggesting to `__user_ptr` in the function that keep `__orig_free`, am I correctly understand that? Well, thus we lack symmetry with rest of `__get_original*` functions.

Reply:
I think I tend to agree with what Konstantin is proposing. Perhaps, we should extrapolate this approach to other places. The code with the changes Konstantin made, the code looks cleaner to me/